File: frontends/admin-ui/src/api/legal.ts
import { request } from "./http";

export type LegalDocument = {
  id: string;
  code: string;
  version: string;
  title: string;
  locale: string;
  effective_from: string;
  status: string;
  content_type: string;
  content: string;
  content_hash: string;
  published_at: string | null;
  created_at: string;
  updated_at: string;
};

export type LegalAcceptance = {
  id: string;
  subject_type: string;
  subject_id: string;
  document_code: string;
  document_version: string;
  document_locale: string;
  accepted_at: string;
  ip: string | null;
  user_agent: string | null;
  acceptance_hash: string;
  signature: Record<string, unknown> | null;
  meta: Record<string, unknown> | null;
};

export async function listLegalDocuments(token: string, params?: Record<string, string>): Promise<LegalDocument[]> {
  const search = new URLSearchParams(params ?? {}).toString();
  const suffix = search ? `?${search}` : "";
  const response = await request<{ items: LegalDocument[] }>(`/admin/legal/documents${suffix}`, { method: "GET" }, token);
  return response.items ?? [];
}

export async function createLegalDocument(token: string, payload: Omit<LegalDocument, "id" | "status" | "content_hash" | "published_at" | "created_at" | "updated_at">) {
  return request<LegalDocument>(
    "/admin/legal/documents",
    { method: "POST", body: JSON.stringify(payload) },
    token,
  );
}

export async function updateLegalDocument(token: string, id: string, payload: Partial<LegalDocument>) {
  return request<LegalDocument>(`/admin/legal/documents/${id}`, { method: "PUT", body: JSON.stringify(payload) }, token);
}

export async function publishLegalDocument(token: string, id: string) {
  return request<LegalDocument>(`/admin/legal/documents/${id}/publish`, { method: "POST" }, token);
}

export async function listLegalAcceptances(token: string, params?: Record<string, string>): Promise<LegalAcceptance[]> {
  const search = new URLSearchParams(params ?? {}).toString();
  const suffix = search ? `?${search}` : "";
  const response = await request<{ items: LegalAcceptance[] }>(
    `/admin/legal/acceptances${suffix}`,
    { method: "GET" },
    token,
  );
  return response.items ?? [];
}

export async function fetchLegalRequired(token: string) {
  return request<{ required: unknown[]; is_blocked: boolean; enabled?: boolean }>("/legal/required", { method: "GET" }, token);
}

export async function acceptLegalDocument(token: string, payload: { code: string; version: string; locale: string }) {
  return request("/legal/accept", { method: "POST", body: JSON.stringify({ ...payload, accepted: true }) }, token);
}

File: platform/processing-core/app/schemas/legal.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class LegalSubject(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: str
    id: str


class LegalRequiredItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str
    title: str
    locale: str
    required_version: str
    published_at: datetime | None
    effective_from: datetime
    content_hash: str
    accepted: bool
    accepted_at: datetime | None


class LegalRequiredResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    subject: LegalSubject
    required: list[LegalRequiredItem]
    is_blocked: bool
    enabled: bool = True


class LegalDocumentResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    code: str
    version: str
    title: str
    locale: str
    effective_from: datetime
    status: str
    content_type: str
    content: str
    content_hash: str
    published_at: datetime | None
    created_at: datetime
    updated_at: datetime


class LegalAcceptRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str
    version: str
    locale: str
    accepted: bool
    signature: dict[str, Any] | None = None
    meta: dict[str, Any] | None = None


class LegalDocumentCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str
    version: str
    title: str
    locale: str = "ru"
    effective_from: datetime
    content_type: str
    content: str


class LegalDocumentUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str
    locale: str
    effective_from: datetime
    content_type: str
    content: str


class LegalDocumentListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[LegalDocumentResponse]


class LegalAcceptanceResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    subject_type: str
    subject_id: str
    document_code: str
    document_version: str
    document_locale: str
    accepted_at: datetime
    ip: str | None
    user_agent: str | None
    acceptance_hash: str
    signature: dict[str, Any] | None
    meta: dict[str, Any] | None


class LegalAcceptanceListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[LegalAcceptanceResponse]


__all__ = [
    "LegalAcceptRequest",
    "LegalAcceptanceListResponse",
    "LegalAcceptanceResponse",
    "LegalDocumentCreateRequest",
    "LegalDocumentListResponse",
    "LegalDocumentResponse",
    "LegalDocumentUpdateRequest",
    "LegalRequiredItem",
    "LegalRequiredResponse",
    "LegalSubject",
]
